lang/thread.md

```diff
@@ -0,0 +1,162 @@
+        let init_frame = StackFrame {
+            func,
+            locals: Map::new(),
+            caller_return_info: None,
+            next_block: func.start,
+            next_stmt: Int::ZERO,
+        };
+
+        Self {
+            state: ThreadState::Enabled,
+            stack: list![init_frame],
+        }
+    }
+}
+```
+
+Some functionality.
+
+```rust
+impl<M: Memory> Thread<M> {
+    fn cur_frame(&self) -> StackFrame<M> {
+        self.stack.last().unwrap()
+    }
+
+    fn mutate_cur_frame<O>(&mut self, f: impl FnOnce(&mut StackFrame<M>) -> O) -> O {
+        if self.stack.is_empty() {
+            panic!("`mutate_cur_frame` called on empty stack!");
+        }
+
+        let last_idx = self.stack.len() - 1;
+        self.stack.mutate_at(last_idx, f)
+    }
+
+    fn mutate_stack<O>(&mut self, f: impl FnOnce(&mut List<StackFrame<M>>) -> O) -> O {
+        f(&mut self.stack)
+    }
+}
+```
+
+## MiniRust Thread Manager
+
+This defines the thread manager. It keeps state about threads during a step execution but also over the whole execution.
+It decides which thread will perform step next.
+
+```rust
+// This is the thread manager it has state about the overall execution
+// as well as the execution of a single step.
+pub struct ThreadManager<M: Memory> {
+    // The set of threads.
+    threads: List<Thread<M>>,
+
+    // This is to keep state during a step.
+    active_thread: Option<ThreadId>,
+}
+```
+
+Some functionality.
+
+```rust
+impl<M: Memory> ThreadManager<M> {
+    pub fn new(func: Function) -> Self {
+        let master = Thread::new(func);
+
+        let mut threads = List::new();
+        threads.push(master);
+
+        Self {
+            threads,
+            active_thread: None,
+        }
+    }
+
+    pub fn new_thread(&mut self, func: Function) -> NdResult<ThreadId> {
+
+        let thread_id = ThreadId::from(self.threads.len());
+
+        self.threads.push(Thread::new(func));
+
+        ret(thread_id)
+    }
+
+    // Called in the beginning of step to decide which thread gets executed next.
+    pub fn schedule(&mut self) -> NdResult<()> {
+        // TODO: Make this just the existing thread_ids.
+        let distr = libspecr::IntDistribution {
+            start: Int::ZERO,
+            end: Int::from(self.threads.len()),
+            divisor: Int::ONE,
+        };
+
+        let thread_id: ThreadId = pick(distr, |id: ThreadId| {
+            let thread = self.threads[id];
+
+            if thread.state != ThreadState::Enabled { return false; }
+
+            true
+        })?;
+
+        self.active_thread = Some(thread_id);
+
+        ret(())
+    }
+
+    pub fn terminate_active_thread(&mut self) -> NdResult {
+        let active = self.active_thread.unwrap();
+
+        if active == 0 {
+            // Consistency with previous system, where the bottom
+            // frame in the stack was never allowed to return.
+            throw_ub!("Master thread terminated.");
+        }
```

Comment:
This is a strange special case... if we need it anyway we might as well throw a nice MachineStop here?